RoboSats/robosats
# Use nostr as cache system

## What does this PR do?

This is a first POC. I'm exploring to use nostr not as primary DB but as a cache system of a decentralized network such as the federation of Robosats, right now a client has to directly connect to every single coordinator book to fetch its specific orders.

 Centralizing this system would create a single point of failure, so I'm exploring a federation cache network, where the full book is aditioanlly stored  and maintained by coordinators, allowing users to initially connect just to 2-3 coordinators. any feedback is welcome 🙂

For now this PR just creates a coordinator relay and sends an event every time a new order is created. The note follows this on going NIP https://github.com/nostr-protocol/nips/pull/1331

As additional win, this will allow Robosats to quickly integrate others' platforms orders with just a single implementation

## Federation cache

Every relay of the federation will be in sync with the others, containing all orders available in the federation:

```mermaid
flowchart TD;
    classDef coordinator fill:#F7931A
    classDef nostr fill:#8e30eb
    subgraph 1
    Coor1:::coordinator-->Relay1;
    end
    subgraph 2
    Coor2:::coordinator-->Relay2;
    end
    subgraph 3
    Coor3:::coordinator-->Relay3;
    end

    Relay1:::nostr<-->Relay2;
    Relay2:::nostr<-->Relay3;
    Relay3:::nostr<-->Relay1;
    Relay2<-->ExternalRelay;

    Relay1-->Client;
    Relay2-->Client;
```
The federation will keep a strong [negentropy](https://github.com/hoytech/negentropy) circle between their relays, but any other external relay will be able to fetch or create new order events. Because it's just cache, we can leave the creation of events open to anyone, if it becomes a spam problem, it can be cleaned up automatically by coordinators.

Clients will always send requests filtering by coordinators' public keys, that way spam or fake orders will never reach to the user.

## How to test

Publish an order. Connect your WS tools to `ws://localhost:12596/nostr` and send:

```
["REQ", "111", {"kinds":[38383]}]
```
You should get in return:

```
["EVENT", "111", <your order note>]
["EOSE","111"]  # End of response
```
Any change on the order should remove previous note and create a new one (identified over time by its `d` tag). To keep privacy, Robosats will only have 2 order statuses: `pending` and `success`.

## Checklist before merging
- [x] Install [pre-commit](https://pre-commit.com/#installation) and initialize it: `pip install pre-commit`, then `pre-commit install`. Pre-commit installs git hooks that automatically check the codebase. If pre-commit fails when you commit your changes, please fix the problems it points out.
